backend/src/AppServer.hs

```diff
@@ -1,38 +1,98 @@
 module AppServer where
 
 import Api (Api)
+import Data.IntMap.Strict qualified as Map
 import Data.Pool (withResource)
-import Data.Time (UTCTime)
 import Data.Time.Clock (getCurrentTime)
 import Data.Validation (Validation (..))
-import Database (insertGameReport, insertPlayerIfNotExists)
+import Database (getLatestRating, insertGameReport, insertPlayerIfNotExists, insertRatingChange)
 import Database.SQLite.Simple (withTransaction)
 import Servant (ServerError (errBody), ServerT, throwError)
 import Servant.Server (err422)
-import Types.Api (GameReport (..))
+import Types.Api (GameReport (..), SubmitGameReportResponse (..))
 import Types.App (AppM, Env (..))
-import Types.DataField (PlayerId)
-import Types.Database (ReadProcessedGameReport (..), WriteProcessedGameReport (..))
+import Types.DataField (Rating, Side (..))
+import Types.Database (ReadProcessedGameReport (..), WriteProcessedGameReport (..), WriteRatingChange (..))
 import Validation (validateReport)
 
-processReport :: UTCTime -> PlayerId -> PlayerId -> GameReport -> WriteProcessedGameReport
-processReport timestamp winnerId loserId (GameReport {..}) =
-  WriteProcessedGameReport
-    { winnerRatingAfter = 0, -- TODO
-      loserRatingAfter = 0, -- TODO
-      ..
-    }
+defaultRating :: Rating
+defaultRating = 500
 
-submitReportHandler :: GameReport -> AppM ReadProcessedGameReport
+ratingThresholds :: IntMap (Rating, Rating)
+ratingThresholds =
+  fromList
+    [ (10, (16, 16)),
+      (33, (15, 17)),
+      (56, (14, 18)),
+      (79, (13, 19)),
+      (102, (12, 20)),
+      (126, (11, 21)),
+      (151, (10, 22)),
+      (178, (9, 23)),
+      (207, (8, 24)),
+      (236, (7, 25)),
+      (270, (6, 26)),
+      (308, (5, 27)),
+      (352, (4, 28)),
+      (409, (3, 29)),
+      (499, (2, 30))
+    ]
+
+maxThreshold :: (Rating, Rating)
+maxThreshold = (1, 31)
+
+ratingAdjustment :: Rating -> Rating -> Rating
+ratingAdjustment winner loser
+  | winner >= loser = smallAdjust
+  | otherwise = bigAdjust
+  where
+    diff = abs (winner - loser)
+    (smallAdjust, bigAdjust) = maybe maxThreshold snd (Map.lookupGE diff ratingThresholds)
```

Comment:
It says a lot about your code readability that this made sense to me! Am I reading it right that the logic is:

1. Get the difference between their ratings and look up the corresponding `(smallAdjust, bigAdjust)` in the `ratingThresholds` map.
2. If the difference is higher than 499, use `(1, 31)` as the adjustment options
3. If the winner was previously rated below the user, apply the `bigAdjust`; otherwise the `smallAdjust`

Reply:
Yep! Exactly what you said. Beating someone higher ranked than you gets you more points.